### ISSUE TITLE
feat: support plugins system

### ISSUE THREAD

**Issue by TuiMao233** [YunYouJun/valaxy#55]
In order to better distinguish the responsibilities, we need to rename the configuration name of the workspace

`valaxy.config.ts` -> `blogs.config.{ts|js|yml|json}`

`vite.config.ts` -> `valaxy.config.{ts|js|yml|json}`

- blogs.config remains the same as before, it is still the various configurations of the blog
- valaxy.config Will be our main configuration file, it can still optionally use `yml|json`, if you want more configuration (functions), user|theme can be converted to `js|ts`

```ts
// valaxy.config.ts
import { defineConfig } from 'valaxy'

const config = defineConfig({
    plugins: [
      // plugins settings
    ],
    vite: {},
    vue: {},
    pages: {},
    components: {},
    markdown: {},
    // ...
})
```

This will be the main configuration of `user` | `theme`, maybe our theme can not even need to be packaged and merged

We focus on plugins configuration

```ts
// valaxy.config.ts
import { defineConfig } from 'valaxy'
import pluginCursorEffects from 'valaxy-plugin-cursor-effects'

const config = defineConfig({
    plugins: [
      // icon settings, 
     ['valaxy-plugin-icons',  { custmos: true, optimize: true }]
     // string form
    'valaxy-plugin-dynamic-title',
    // function form
    pluginCursorEffects({ size: 2, zIndex: 10 })
    ]
})
```

for plugins that have been enabled by default, we regard it as a reconfiguration of the plugin, such as `valaxy-plugin-icons`. For some specific plugins, we will process them first and finally get the information of the entire plugin

**Comment by TuiMao233** [comment]
Internal structure of the plugin

- main.ts
- valaxy.config.ts
- package.json

main.ts is similar to the entry of the vue project, he is like this

```ts
import { defineAppSetup } from 'valaxy'

import "./style.scss"
export default defineAppSetup(({ app }) => {

})
```

And valve.config.ts is called as part of the node plugin, which is consistent with the user's

Plugins cannot be referenced inside plugins to avoid multi-level plugins references
```ts
// valaxy.config.ts
import { defineConfig } from 'valaxy'

const config = defineConfig({
    vite: {},
    vue: {},
    pages: {},
    components: {},
    markdown: {},
    // ...
})
```

**Comment by YunYouJun** [comment]
Considering that the theme needs to export the `ThemeConfig` and other type, It may always need to be compiled.
> Certainly consider using `jiti` and `d.ts`.

I think I see what you mean.
But I don't think you need to rename the configuration file.
Plugins can still be placed in `vite.valaxy.plugins` or (valaxy.config.ts).

Just like old VuePress, some of its plugins seem redundant to me.
![image](https://user-images.githubusercontent.com/25154432/179153435-6c6d3cb5-238f-46cd-bab1-467c0e8e6e0e.png)
`plugin-pwa` You can simply use the existing `vite-plugin-pwa` directly.
`backToTop` should be built in (hook) and styles implemented by the theme.

It seems to me that vitepress, as vuepress's successor, is trying to do just that.
More plug-ins don't seem to make it easier, they seem to make it more complicated.
I think that if a plug-in is needed, it only needs to provide what a Vite plug-in can't.

**Comment by YunYouJun** [comment]
@jw-foss What do you think?

**Comment by TuiMao233** [comment]
No, we shouldn't use vite.config.ts, it's gradually out of scope, and honestly vite it should be clean

I did this renaming in another branch https://github.com/TuiMao233/valaxy

**Comment by TuiMao233** [comment]
In the end, our configuration should have a clear division of labor and a clean

- blog.config
- valaxy.config
- vite.config

**Comment by TuiMao233** [comment]
> More plug-ins don't seem to make it easier, they seem to make it more complicated.

In fact, the use of plugins is determined by the user, you cannot make the decision for them

A long time ago, as a beginner, it was really great to see that vuepress has so many plugins support, because I can have a lot of coveted features without writing any code.

**Comment by LoTwT** [comment]
> Internal structure of the plugin
> 
> * main.ts
> * index.ts
> * package.json
> 
> main.ts is similar to the entry of the vue project, he is not necessary, he is like this
> 
> ```ts
> import BackToTop from './BackToTop.vue'
> export default ({ app }) => {
>   app.component('BackToTop', BackToTop)
> }
> ```
> 
> And index.ts is called as part of the node plugin, which is basically the same as the user's
> 
> Plugins cannot be referenced inside plugins to avoid multi-level plugins references
> 
> ```ts
> // index.ts
> import { definePlugin } from 'valaxy'
> import path from 'path'
> 
> const config = definePlugin((options) =>{
>   return {
>       // --- plugin unique config ---
>       name: 'valaxy-plugin-custmo-icons',
>       enhanceAppFiles: [
>         path.resolve(__dirname, './main.ts')
>       ],
>       globalUIComponents: 'BackToTop',
>       // --- plugin unique config ---
>       vite: {},
>       vue: {},
>       pages: {},
>       components: {},
>       markdown: {},
>       // ...
>   }
> })
> 
> export default config
> ```
> 
> As for layout and components, I'm not sure whether they should be supported, this has been implemented in the theme, and this will bring more processing, I think it can be ignored in the short term

Do you mean a valaxy plugin needs to support `client` and `node` at the same time ? I think they should be separate, following Separation of concerns.

**Comment by YunYouJun** [comment]
New addon ideas.

- App.vue
- index.ts
- package.json

`package.json` has `global` property to determine whether global mounted.

And `valaxy` merge `addon/App.vue` and `user/App.vue` by `global`.

**Comment by TuiMao233** [comment]
> 三个配置文件太多了。他们可能会造成更多的精神负担。我认为一两个不同的属性就足够了。我认为用户总是希望使用功能而不是编写配置文件。在一个配置文件中，按照直觉，有几个属性适合目的。

好的，我们来分析一下配置

首先，我们得知道我们的配置是为谁服务的
1. 用户
2. 主题
3. 插件
4. 外来者

考虑到用户的切换成本，这也是为什么我们需要分工明确，我们得知道这个概念

例如我是一个 vite project 项目的开发者，我希望拥有一个博客，但我也希望我可以从我项目中的 vite 配置得到应用

那么会发生以下的转变过程

`vite.config.ts` > `valaxy/vite.config.ts`

假如我是一个用户，我为博客新增加了很多功能，希望能为此做出一些贡献，希望能将现在的配置 转换成 theme 

那么会发生以下的转变过程

`valaxy.config.ts` > `theme/valaxy.config.ts`

那么在这个过程里面，`site.config.ts` 依旧是用户独有的

**Comment by TuiMao233** [comment]
以及用户依然可以只使用 `site.config.ts`，它没有任何影响，依旧享受 valaxy 生态带来的痛快

当用户需要更强大的功能，我们 `valaxy` 就能提供它想要的任何细节，例如提供一个 `valaxy.config` 控制 `valaxy` 内部的所有细节

当用户希望分享主题，它只需要将配置转移到应该新的 theme project 中

**Comment by YunYouJun** [comment]
> Plugins cannot be referenced inside plugins to avoid multi-level plugins references

If for that purpose, `plugins` should be placed in `site.config.ts`.
Theme do not need plugins too.

**Comment by jw-foss** [comment]
As for me, **3** configuration files make the most sense. Here are the reasons:

1. Most people **DO NOT** need that many configurations, since we provide out of box blogging functionalities.
2. For user who wants to make their own `theme` or `add-ons`, I think they will need to configure everything themselves.
3. I agree with @TuiMao233 on not touching anything with `vite.config` since it is the base of `Valaxy` and it should be left alone. Of course the users may want some additional functionalities on `vite`, they can do that by modifying the `vite.config` file themselves.

**Comment by jw-foss** [comment]
As for the `add-on` itself. For me personally, I think as a blogging framework, the add-on itself should be something like a component, if we want functionalities from `vite`, we simply update `vite.config` file.

With this kind of mindset, we should consider `Theme` as a superset of `add-on`. 

`Theme` should be a collection of components that we want to customize based on the template. 

`Add-on` should be a component that is scattered and can be referenced in the blogging framework.

Consider the `Click` component in `Slidev`, `Add-on` can be implemented as `Click` and can be referenced globally.

**Comment by jw-foss** [comment]
Let me know what you guys think 😉

**Comment by TuiMao233** [comment]
> If for that purpose, plugins should be placed in site.config.ts.

In fact I think that theme can provide users with plugins, which is great

some excellent themes come with a lot of plugins, we can't limit the theme for this reason

https://shoka.lostyu.me/computer-science/note/theme-shoka-doc/dependents/

We should focus on addons

**Comment by TuiMao233** [comment]
> With this kind of mindset, we should consider `Theme` as a superset of `add-on`.

add-on if used as a component, it has only one, not multiple components

The add-on has two sides, it has the function on the global UI and also has the function of the actual plugin, i.e. node

**Comment by TuiMao233** [comment]
> In conclusion: 3 configurations make sense, but we can simplify it into 2, and for vite.config should stay intact and we only consume it.

In fact, `valaxy.config` should be a separate configuration, and from the point of view of `user`, `theme`, they should be consistent, but in fact, users may not need `valaxy.config`, and Themes don't need `site.config` at all
